File: src/test/java/application/browser/CardHelper.java
package application.browser;

import com.matag.cards.Card;
import com.matag.cards.ability.type.AbilityType;
import com.matag.player.PlayerType;
import lombok.AllArgsConstructor;
import org.openqa.selenium.*;

import java.util.List;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.openqa.selenium.support.ui.ExpectedConditions.*;

@AllArgsConstructor
public class CardHelper {
  private final WebElement webElement;
  private final MatagBrowser matagBrowser;

  public static List<String> cardNames(Card... cards) {
    return Stream.of(cards).map(Card::getName).sorted().collect(toList());
  }

  public static List<String> cardNames(WebElement element) {
    var cardElements = element.findElements(By.className("card"));
    return cardElements.stream()
      .map(cardElement -> cardElement.getAttribute("aria-label"))
      .sorted()
      .collect(toList());
  }

  public void click() {
    try {
      webElement.click();
    } catch (ElementClickInterceptedException e) {
      matagBrowser.getJavascriptExecutor().executeScript("arguments[0].click()", webElement);
    }
  }

  public void tap() {
    playerHasPriority();
    click();
    isFrontendTapped();
  }

  public void select() {
    playerHasPriority();
    click();
    isSelected();
  }

  public void target() {
    playerHasPriority();
    click();
    isTargeted();
  }

  public void declareAsAttacker() {
    declareAsAttackerOrBlocker();
  }

  public void declareAsBlocker() {
    declareAsAttackerOrBlocker();
  }

  public String getCardId() {
    return webElement.getAttribute("id");
  }

  public int getCardIdNumeric() {
    return Integer.parseInt(getCardId().split("-")[1]);
  }

  public void isFrontendTapped() {
    hasClass("frontend-tapped");
  }

  public void isNotFrontendTapped() {
    doesNotHaveClass("frontend-tapped");
  }

  public void isTapped() {
    hasClass("tapped");
    doesNotHaveClass("frontend-tapped");
  }

  public void isTappedDoesNotUntapNextTurn() {
    hasClass("tapped-does-not-untap-next-turn");
  }

  public void isNotTapped() {
    doesNotHaveClass("tapped");
  }

  public void isTargeted() {
    hasClass("targeted");
  }

  public void isSelected() {
    hasClass("selected");
  }

  public void isNotSelected() {
    doesNotHaveClass("selected");
  }

  public void hasSummoningSickness() {
    matagBrowser.wait(presenceOfElementLocated(cardCssSelector(".summoning-sickness")));
  }

  public void doesNotHaveSummoningSickness() {
    waitForAbsenceOfElement(cardCssSelector(".summoning-sickness"));
  }

  public void hasFlying() {
    hasClass("flying");
  }

  public void hasDamage(int damage) {
    matagBrowser.wait(textToBe(cardCssSelector(".damage"), String.valueOf(damage)));
  }

  public void doesNotHaveDamage() {
    waitForAbsenceOfElement(cardCssSelector(".damage"));
  }

  public void hasPowerAndToughness(String powerAndToughness) {
    matagBrowser.wait(textToBe(cardCssSelector(".power-toughness"), String.valueOf(powerAndToughness)));
  }

  public void parentHasStyle(String style) {
    var parent = webElement.findElement(By.xpath("./.."));
    assertThat(parent.getAttribute("style")).contains(style);
  }

  private void hasClass(String classValue) {
    matagBrowser.wait(attributeContains(By.id(getCardId()), "class", classValue));
  }

  private void doesNotHaveClass(String classValue) {
    matagBrowser.wait(not(attributeContains(By.id(getCardId()), "class", classValue)));
  }

  private By cardCssSelector(String cssSelector) {
    return By.cssSelector("#" + getCardId() + " " + cssSelector);
  }

  private void waitForAbsenceOfElement(By locator) {
    try {
      matagBrowser.findElement(locator);
      matagBrowser.wait(invisibilityOfElementLocated(locator));

    } catch (TimeoutException | NotFoundException e) {
      System.out.println("Element " + locator + " is not present at all. That's okay.");
    }
  }

  private void playerHasPriority() {
    new PhaseHelper(matagBrowser).isPriority(PlayerType.PLAYER);
  }

  private void declareAsAttackerOrBlocker() {
    playerHasPriority();
    var cardId = getCardId();
    click();
    matagBrowser.wait(presenceOfElementLocated(By.cssSelector(".combat-line #" + cardId)));
  }

  public void hasPlus1Counters(int counters) {
    matagBrowser.wait(textToBe(cardCssSelector(".plus-1-counter"), String.valueOf(counters)));
  }

  public void hasMinus1Counters(int counters) {
    matagBrowser.wait(textToBe(cardCssSelector(".minus-1-counter"), String.valueOf(counters)));
  }

  public void hasKeywordCounters(AbilityType keywordAbility) {
    matagBrowser.wait(attributeContains(By.cssSelector("#" + this.getCardId() + " .keyword-counter"), "title", keywordAbility.name().toLowerCase()));
  }
}

File: src/test/kotlin/application/browser/CardHelper.kt
package application.browser

import com.matag.cards.Card
import com.matag.cards.ability.type.AbilityType
import com.matag.player.PlayerType
import org.assertj.core.api.Assertions
import org.openqa.selenium.*
import org.openqa.selenium.support.ui.ExpectedConditions
import java.util.*

class CardHelper(
    private val webElement: WebElement,
    private val matagBrowser: MatagBrowser
) {
    fun click() {
        try {
            webElement.click()
        } catch (_: ElementClickInterceptedException) {
            matagBrowser.getJavascriptExecutor().executeScript("arguments[0].click()", webElement)
        }
    }

    fun tap() {
        playerHasPriority()
        click()
        isFrontendTapped()
    }

    fun select() {
        playerHasPriority()
        click()
        isSelected()
    }

    fun target() {
        playerHasPriority()
        click()
        isTargeted()
    }

    fun declareAsAttacker() {
        declareAsAttackerOrBlocker()
    }

    fun declareAsBlocker() {
        declareAsAttackerOrBlocker()
    }

    fun getCardId(): String =
        webElement.getAttribute("id")!!

    fun getCardIdNumeric(): Int =
        getCardId().split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1].toInt()

    fun isFrontendTapped() =
        hasClass("frontend-tapped")

    fun isNotFrontendTapped() =
        doesNotHaveClass("frontend-tapped")

    fun isTapped() =
         hasClass("tapped") && doesNotHaveClass("frontend-tapped")

    fun isTappedDoesNotUntapNextTurn() =
        hasClass("tapped-does-not-untap-next-turn")

    fun isNotTapped() =
        doesNotHaveClass("tapped")

    fun isTargeted() =
        hasClass("targeted")

    fun isSelected() =
        hasClass("selected")

    fun isNotSelected() =
        doesNotHaveClass("selected")

    fun hasSummoningSickness() =
        matagBrowser.wait(ExpectedConditions.presenceOfElementLocated(cardCssSelector(".summoning-sickness")))

    fun doesNotHaveSummoningSickness() =
        waitForAbsenceOfElement(cardCssSelector(".summoning-sickness"))

    fun hasFlying() =
        hasClass("flying")

    fun hasDamage(damage: Int) =
        matagBrowser.wait(ExpectedConditions.textToBe(cardCssSelector(".damage"), damage.toString()))

    fun doesNotHaveDamage() =
        waitForAbsenceOfElement(cardCssSelector(".damage"))

    fun hasPowerAndToughness(powerAndToughness: String) =
        matagBrowser.wait(
            ExpectedConditions.textToBe(
                cardCssSelector(".power-toughness"),
                powerAndToughness
            )
        )

    fun parentHasStyle(style: String?) {
        val parent = webElement.findElement(By.xpath("./.."))
        Assertions.assertThat(parent.getAttribute("style")).contains(style)
    }

    private fun hasClass(classValue: String) =
        matagBrowser.wait(ExpectedConditions.attributeContains(By.id(getCardId()), "class", classValue)).let { true }

    private fun doesNotHaveClass(classValue: String) =
        matagBrowser.wait(ExpectedConditions.not(ExpectedConditions.attributeContains(By.id(getCardId()), "class", classValue))).let { true }

    private fun cardCssSelector(cssSelector: String?): By =
        By.cssSelector("#" + getCardId() + " " + cssSelector)

    private fun waitForAbsenceOfElement(locator: By) {
        try {
            matagBrowser.findElement(locator)
            matagBrowser.wait(ExpectedConditions.invisibilityOfElementLocated(locator))
        } catch (_: TimeoutException) {
            println("Element $locator is not present at all. That's okay.")
        } catch (_: NotFoundException) {
            println("Element $locator is not present at all. That's okay.")
        }
    }

    private fun playerHasPriority() =
        PhaseHelper(matagBrowser).isPriority(PlayerType.PLAYER)

    private fun declareAsAttackerOrBlocker() {
        playerHasPriority()
        val cardId = getCardId()
        click()
        matagBrowser.wait(ExpectedConditions.presenceOfElementLocated(By.cssSelector(".combat-line #$cardId")))
    }

    fun hasPlus1Counters(counters: Int) =
        matagBrowser.wait(ExpectedConditions.textToBe(cardCssSelector(".plus-1-counter"), counters.toString()))

    fun hasMinus1Counters(counters: Int) =
        matagBrowser.wait(ExpectedConditions.textToBe(cardCssSelector(".minus-1-counter"), counters.toString()))

    fun hasKeywordCounters(keywordAbility: AbilityType) =
        matagBrowser.wait(
            ExpectedConditions.attributeContains(
                By.cssSelector("#" + getCardId() + " .keyword-counter"),
                "title",
                keywordAbility.name.lowercase(Locale.getDefault())
            )
        )

    companion object {
        fun cardNames(cards: List<Card>): List<String> =
            cards.map { it.name }.sorted()

        fun cardNames(element: WebElement): List<String> {
            val cardElements = element.findElements(By.className("card"))
            return cardElements
                .map { it.getAttribute("aria-label")!! }
                .sorted()
        }
    }
}
